### PR TITLE
ENH add partial_fit for DecisionTreeClassifier

### PR DESCRIPTION
**Pull request by PSSF23**
@adam2392 I have revamped the `update` cython function into `build` as we discussed in #35. Right now I only modified `DepthFirstTreeBuilder` as there's no way to control max depth in streaming trees.

### CODE REVIEW

File: sklearn/tree/_tree.pyx [hunk 1]
```diff
@@ -896,6 +1023,56 @@ cdef class BaseTree:
...
+            node.right_child = _TREE_LEAF
+            node.feature = _TREE_UNDEFINED
+            node.threshold = _TREE_UNDEFINED
+        else:
+            node.feature = split_node.feature
+            node.threshold = split_node.threshold
+
+        return node_id
+
+    cpdef np.ndarray predict(self, object X):
```
Comment (adam2392) [inline sklearn/tree/_tree.pyx:1068]: `BaseTree` doesn't need `predict` since `BaseTree` can be used for unsupervised tree. I think `predict` is defined in `Tree(BaseTree)` class.

File: sklearn/tree/_tree.pyx [hunk 2]
```diff
@@ -896,6 +1023,56 @@ cdef class BaseTree:
...
+        SIZE_t parent,
+        bint is_left,
+        bint is_leaf,
+        SplitRecord* split_node,
+        double impurity,
+        SIZE_t n_node_samples,
+        double weighted_n_node_samples,
+        unsigned char missing_go_to_left
+    ) except -1 nogil:
+        """Update a node on the tree.
```
Comment (adam2392) [inline sklearn/tree/_tree.pyx:1037]: To make this work for oblique splits in scikit-tree, I guess we would need to also update `proj_vec_indices` and `proj_vec_weights`(?) Nothing to do here I think(?)

File: sklearn/tree/_tree.pyx [hunk 3]
```diff
@@ -206,18 +286,22 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
         cdef SIZE_t min_samples_split = self.min_samples_split
         cdef double min_impurity_decrease = self.min_impurity_decrease
 
-        # Recursive partition (without actual recursion)
-        splitter.init(X, y, sample_weight, missing_values_in_feature_mask)
+        if self.initial_roots:
+            pass
+        else:
```
Comment (adam2392) [inline sklearn/tree/_tree.pyx:291]: why not `if not self.initial_roots:`?

**Review by adam2392 — COMMENTED** [review]
I left very high-level brief skim notes.

Can you remind me again what is the difference between adding a node and updating a node and why we need separate functions? They look the same kind of. 

I can help marathon this in next week perhaps and help you run some exps to improve the implementation **fingers crossed if I have time**. Might even be helpful for your paper ;)

**Comment by PSSF23** [comment]
Thanks for the notes! I'm still cleaning up the merging conflicts and will get back to your reviews once I have the code running again.

File: sklearn/tree/_tree.pyx (hunk 2 above)
Comment (PSSF23) [reply sklearn/tree/_tree.pyx:1032]: Are those variables stored with the node?

File: sklearn/tree/_tree.pyx (hunk 3 above)
Comment (PSSF23) [reply sklearn/tree/_tree.pyx:288]: modified

**Review by PSSF23 — COMMENTED** [review]
> @adam2392 In `_update_node` the tree finds the node id of the current node instead of adding a new one. In my opinion, combining the two functions would not necessarily simplify the code as there would be many if statements..

But `_add_node` looks almost exactly the same. Wouldn't it be easier to add an `overwrite=False` to the `_add_node` function?

```
if node_id >= self.capacity:
            if self._resize_c() != 0:
                return INTPTR_MAX

        cdef Node* node = &self.nodes[node_id]
        node.impurity = impurity
        node.n_node_samples = n_node_samples
        node.weighted_n_node_samples = weighted_n_node_samples

        if parent != _TREE_UNDEFINED:
            if is_left:
                self.nodes[parent].left_child = node_id
            else:
                self.nodes[parent].right_child = node_id

        if is_leaf:
            if self._set_leaf_node(split_node, node) != 1:
                with gil:
                    raise RuntimeError
        else:
            if self._set_split_node(split_node, node) != 1:
                with gil:
                    raise RuntimeError
            node.missing_go_to_left = missing_go_to_left

        self.node_count += 1

        return node_id
```

File: sklearn/tree/_tree.pxd [hunk 4]
```diff
@@ -148,9 +174,19 @@ cdef class TreeBuilder:
     cdef double min_weight_leaf         # Minimum weight in a leaf
     cdef SIZE_t max_depth               # Maximal tree depth
     cdef double min_impurity_decrease   # Impurity threshold for early stopping
+    cdef object initial_roots
```
Comment (adam2392) [inline sklearn/tree/_tree.pxd:177]: Add a in-line comment for what this is meant to hold.

File: sklearn/tree/_tree.pyx [hunk 5]
```diff
@@ -296,10 +399,17 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
...
+                    if parent in self.initial_roots:
+                        node_id = tree._update_node(parent, is_left, is_leaf,
+                                                    split_ptr, impurity, n_node_samples,
+                                                    weighted_n_node_samples,
+                                                    split.missing_go_to_left)
+                    else:
+                        node_id = tree._add_node(parent, is_left, is_leaf,
+                                                 split_ptr, impurity, n_node_samples,
+                                                 weighted_n_node_samples,
+                                                 split.missing_go_to_left)
```
Comment (adam2392) [inline sklearn/tree/_tree.pyx:412]: This will significantly slow down the tree building because you have to access the GIL again.

If this if parent in self.initial_roots check is necessary, then I think we should just build a C++ hashmap instead, so we don't need the GIL

Comment (PSSF23) [reply sklearn/tree/_tree.pyx:412]: It's an annoying thing as we have node splitting and node updating mixed together. I agree that it's time consuming, and will find a way to simplify it. Which way you think is more applicable, separating the node updates or pushing an indicator onto the stack?

Comment (adam2392) [reply sklearn/tree/_tree.pyx:412]: Indicator seems reasonable?

Alternatively, why not just make `initial_roots` a hashmap? Using chatgpt, the code seems not too bad:

```
# hashmap_cython.pyx

# Import necessary Cython headers
from libcpp.unordered_map cimport unordered_map

# myclass is the treebuilder
cdef class MyClass:
    cdef unordered_map[int, int] map

    def __cinit__(self):
        self.map = unordered_map[int, int]()

    cpdef int insert(self, int key, int value) nogil:
        # Use NOGIL to allow direct access to the unordered_map
        self.map[key] = value
        return 1

    cpdef int find(self, int key) nogil:
        # Use NOGIL to allow direct access to the unordered_map
        try:
            return self.map.at(key)
        except KeyError:
            return -1
```

Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: It seems very complicated to use tuple as hashmap keys, so I separated the node updates and node additions.

Comment (adam2392) [reply sklearn/tree/_tree.pyx]: What is initial roots storing? Hash map of node id as key to what as value?

Comment (adam2392) [reply sklearn/tree/_tree.pyx]: If that is the case, does it makes sense to just refactor it then into:

```
{
     'parent_node_id': {
             'is_left_child': True/False,
             'n_samples': n_samples,
             'depth': depth
     }
}
```

Would this be possible w/o breaking the existing code? I'm just trying to limit code duplication in the fork because otw every time we sync upstream changes, it gets more complex.

Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: I'll get back to this soon. Just realized because merging `submodulev2` changed the package name to `scikit-learn-tree`, I was not testing the updated package.

Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: No, it defaulted to the `scikit-learn` fork I have before the pr.

Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: So you now prefer the original `update` method instead of integrating the function to `build`?

Comment (adam2392) [reply sklearn/tree/_tree.pyx]: Just thinking out diff approaches. If the code as is works now then it's just ideas on how to simplify the implementation so we don't have maintainence work constantly

Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: One advantage of the `update` method is that it would not be affected by any upstream changes.

File: sklearn/model_selection/_split.py [hunk 6]
```diff
@@ -2673,7 +2673,7 @@ def _pprint(params, offset=0, printer=repr):
     this_line_length = offset
     line_sep = ",\n" + (1 + offset // 2) * " "
     for i, (k, v) in enumerate(sorted(params.items())):
-        if type(v) is float:
+        if isinstance(v) is float:
```
Comment (adam2392) [inline sklearn/model_selection/_split.py:2676]: Fluff?

File: sklearn/metrics/tests/test_classification.py [hunk 7]
```diff
@@ -159,10 +159,10 @@ def test_classification_report_dictionary_output():
...
                 assert_almost_equal(expected_report[key][metric], report[key][metric])
 
-    assert type(expected_report["setosa"]["precision"]) == float
-    assert type(expected_report["macro avg"]["precision"]) == float
-    assert type(expected_report["setosa"]["support"]) == int
-    assert type(expected_report["macro avg"]["support"]) == int
+    assert isinstance(expected_report["setosa"]["precision"]) == float
+    assert isinstance(expected_report["macro avg"]["precision"]) == float
+    assert isinstance(expected_report["setosa"]["support"]) == int
+    assert isinstance(expected_report["macro avg"]["support"]) == int
```
Comment (adam2392) [inline sklearn/metrics/tests/test_classification.py:165]: Is this needed?

File: sklearn/tree/_tree.pyx (hunk 5 above)
Comment (adam2392) [reply sklearn/tree/_tree.pyx]: True. The issue is that will you be able to account for monotonic constraints and any other added features easily tho?

File: sklearn/metrics/tests/test_classification.py (hunk 7 above)
Comment (PSSF23) [reply sklearn/metrics/tests/test_classification.py:165]: Yes it's one of the build errors

File: sklearn/tree/_tree.pyx (hunk 5 above)
Comment (PSSF23) [reply sklearn/tree/_tree.pyx]: As I already implemented the initial roots method I would stick with it for now. It seems that the errors come from `fit` underperforming, which my code should have no effects on. I'll compare it with the sklearn code tomorrow and see if I can resolve it. It is affecting both the update method and the build/roots method.

File: sklearn/tree/_tree.pyx (hunk 2 above)
Comment (adam2392) [reply sklearn/tree/_tree.pyx:1153]: Yeah in downstream oblique trees, we additionally handle a projection vector of indices and weights.

To enable OF/MORF to have the partial_fit via `update` function, we will just need to perhaps override the `update_node` function to handle additionally the `proj_vec_indices/weights`. Then the `TreeBuilder.update` function can be used for any tree.

File: sklearn/tree/_tree.pyx (hunk 5 above)
Comment (adam2392) [reply sklearn/tree/_tree.pyx]: Let's see if https://github.com/neurodata/scikit-learn/pull/52 will fix it
